code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java

```diff
@@ -0,0 +1,140 @@
+package io.cattle.platform.iaas.api.auth.github;
+
+import io.cattle.platform.api.resource.jooq.AbstractJooqResourceManager;
+import io.cattle.platform.core.constants.AccountConstants;
+import io.cattle.platform.core.model.Account;
+import io.cattle.platform.iaas.api.auth.dao.AuthDao;
+import io.cattle.platform.iaas.api.auth.github.resource.GithubAccountInfo;
+import io.cattle.platform.json.JsonMapper;
+import io.cattle.platform.util.type.CollectionUtils;
+import io.github.ibuildthecloud.gdapi.context.ApiContext;
+import io.github.ibuildthecloud.gdapi.exception.ClientVisibleException;
+import io.github.ibuildthecloud.gdapi.factory.SchemaFactory;
+import io.github.ibuildthecloud.gdapi.model.ListOptions;
+import io.github.ibuildthecloud.gdapi.request.ApiRequest;
+import io.github.ibuildthecloud.gdapi.util.ResponseCodes;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class ProjectResourceManager extends AbstractJooqResourceManager {
+
+    private static final String GITHUB_EXTERNAL_TYPE = "github";
+    private static final String TEAM_SCOPE = "project:github_team";
+    private static final String ORG_SCOPE = "project:github_org";
+    private static final String USER_SCOPE = "project:github_user";
+    private static final String AUTH = "Authorization";
+    private static final String EXTERNAL_ID = "externalId";
+    private static final String EXTERNAL_ID_TYPE = "externalIdType";
+
+    JsonMapper jsonMapper;
+    GithubUtils githubUtils;
+    GithubClient githubClient;
+    AuthDao authDao;
+
+    @Override
+    public String[] getTypes() {
+        return new String[] { "project" };
+    }
+
+    @Override
+    public Class<?>[] getTypeClasses() {
+        return new Class[] {};
+    }
+
+    @Override
+    public Object listInternal(SchemaFactory schemaFactory, String type, Map<Object, Object> criteria, ListOptions options) {
+        HttpServletRequest request = ApiContext.getContext().getApiRequest().getServletContext().getRequest();
+        String token = request.getHeader(AUTH);
+        if (StringUtils.isEmpty(token)) {
+            return null;
+        }
+        Pair<String, Pair<List<String>, List<String>>> ids = githubUtils.validateAndFetchAccesibleIdsFromToken(token);
```

Comment:
Following my comment before, Pairs just obfuscate what the code is doing.  It's the different between i[0] and i.accountId.  One of those is quite obvious.